kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandKeyManager.java

```diff
@@ -76,32 +81,27 @@ public InBandKeyManager(Kms<K, E> kms,
         this.dekTtlNanos = 5_000_000_000L;
         this.maxEncryptionsPerDek = maxEncryptionsPerDek;
         // TODO This ^^ must be > the maximum size of a batch to avoid an infinite loop
-        this.keyContextCache = new ConcurrentHashMap<>();
+        this.keyContextCache = Caffeine.newBuilder()
+                .buildAsync((key, executor) -> makeKeyContext(key));
         this.decryptorCache = new ConcurrentHashMap<>();
         this.encryptionVersion = EncryptionVersion.V1; // TODO read from config
         this.encryptionHeader = new Header[]{ new RecordHeader(ENCRYPTION_HEADER_NAME, new byte[]{ encryptionVersion.code() }) };
     }
 
-    private CompletionStage<KeyContext> getKeyContext(K key,
-                                                      Supplier<CompletionStage<KeyContext>> valueSupplier) {
-        return keyContextCache.compute(key, (k, v) -> {
-            if (v == null) {
-                return valueSupplier.get();
-                // TODO what happens if the CS complete exceptionally
-                // TODO what happens if the CS doesn't complete at all in a reasonably time frame?
-            }
-            else {
-                return v;
-            }
-        });
+    private CompletionStage<KeyContext> currentDekContext(@NonNull K kekId) {
+        // todo should we add some scheduled timeout as well? or should we rely on the KMS to timeout appropriately.
+        return keyContextCache.get(kekId);
     }
 
-    private CompletionStage<KeyContext> currentDekContext(@NonNull K kekId) {
-        return getKeyContext(kekId, makeKeyContext(kekId));
+    private CompletableFuture<KeyContext> makeKeyContext(@NonNull K kekId) {
+        return attemptMakeDekContext(kekId, 0);
     }
 
-    private Supplier<CompletionStage<KeyContext>> makeKeyContext(@NonNull K kekId) {
-        return () -> kms.generateDekPair(kekId)
+    private CompletableFuture<KeyContext> attemptMakeDekContext(@NonNull K kekId, int attempt) {
+        if (attempt == MAX_ATTEMPTS) {
```

Comment:
I think this should probably be a little more defensive:
```suggestion
        if (attempt >= MAX_ATTEMPTS) {
```
Just in cases it gets called in an unexpected fashion.